src/daft-catalog/src/lib.rs

```diff
@@ -0,0 +1,136 @@
+mod data_catalog;
+mod data_catalog_table;
+pub mod errors;
+
+// Export public-facing traits
+use std::{collections::HashMap, default, sync::Arc};
+
+use daft_logical_plan::LogicalPlanBuilder;
+pub use data_catalog::DataCatalog;
+pub use data_catalog_table::DataCatalogTable;
+
+#[cfg(feature = "python")]
+pub mod python;
+
+use errors::Error;
+
+pub mod global_catalog {
+    use std::sync::{Arc, Mutex};
+
+    use lazy_static::lazy_static;
+
+    use crate::{DaftMetaCatalog, DataCatalog};
+
+    lazy_static! {
+        pub(crate) static ref GLOBAL_DAFT_META_CATALOG: Mutex<DaftMetaCatalog> =
+            Mutex::new(DaftMetaCatalog::new_from_env());
+    }
+
+    /// Register a DataCatalog with the global DaftMetaCatalog
+    pub fn register_catalog(catalog: Arc<dyn DataCatalog>, name: Option<&str>) {
+        GLOBAL_DAFT_META_CATALOG
+            .lock()
+            .unwrap()
+            .register_catalog(catalog, name);
+    }
+
+    /// Unregisters a catalog with the global DaftMetaCatalog
+    pub fn unregister_catalog(name: Option<&str>) -> bool {
+        GLOBAL_DAFT_META_CATALOG
+            .lock()
+            .unwrap()
+            .unregister_catalog(name)
+    }
+}
+
+/// Name of the default catalog
+static DEFAULT_CATALOG_NAME: &str = "default";
+
+/// The [`DaftMetaCatalog`] is a catalog of [`DataCatalog`] implementations
+///
+/// Users of Daft can register various [`DataCatalog`] with Daft, enabling
+/// discovery of tables across various [`DataCatalog`] implementations.
+pub struct DaftMetaCatalog {
```

Comment:
Could you elaborate on what you mean by `schema`? Are you referring to the schemas of each table (e.g. datatypes and field names), or the hierarchical naming of tables (e.g. `catalog.schema.table`)?

If it's the latter, I chose to not support any opinionated naming here beyond just catalog name, and then within each catalog just a table name. This is because every catalog has different naming conventions (e.g. AWS Glue tables are nested under the hierarchy of `database.table`, whereas Iceberg would have `namespace.table`) and I didn't think that it would be a good idea for us to try and generalize that

Reply:
I was referring to the latter `catalog.schema.table`. the docstring [here](https://github.com/Eventual-Inc/Daft/pull/3036#discussion_r1842811725) implies that we support not only `<scope>.<table>` (e.g. catalog.table), but also `<scope1>.<scope2>.<table>`. 

I think it'd make sense to dump all catalogs that only have a single scope `<scope>.<table>` inside the `default` catalog (e.g. `default.something.table`) This would likely give us the most flexibility long term, and prevent us from having to do a potentially large refactor later on once we need to support fully qualified `<scope1>.<scope2>.<table>`.

Reply:
Clarified offline:

1. We allow users to register multiple named catalogs. Also we support a default catalog.
2. Users then refer to tables in Daft via: `catalog_name.*`, where `*` is a catalog-dependent string identifier (e.g. Glue, Iceberg etc have their own formats for table names)
3. If users provide a name like `default.*`, or just `*` without the catalog name, we will attempt to resolve the name against our default catalog.